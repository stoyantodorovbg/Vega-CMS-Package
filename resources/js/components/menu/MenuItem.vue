<template>
    <div :class="menuItemData.classes"
         :style="menuItemData.styles"
    >
        <a v-if="menuItemData.title && menuItemData.title.status"
                :href="$helpers.adminUrlPrefix($store.getters.locale) + menuItemData.url"
                :class="menuItemData.title.classes"
                :style="menuItemData.title.styles"
            >
            {{ menuItemData.title.text }}
        </a>
        <div v-if="menuItemData.description && menuItemData.description.status"
             :class="menuItemData.description.classes"
             :style="menuItemData.description.styles"
        >
            {{ menuItemData.description.text }}
        </div>
        <menu-items-container
            v-for="menuItem in menuItemData.menuItems"
            :key="menuItem.id"
            :menuItemData="menuItem"
        ></menu-items-container>
    </div>
</template>

<script>
    export default {
        name: 'menu-item',

        components: {
            MenuItemsContainer: () => import('./MenuItemsContainer.vue'),
        },

        props: {
            menuItemData: {
                type: Object,
                default: function () {
                    return {
                        id: 1,
                        classes: 'text-uppercase',
                        url: '/',
                        title: {
                            classes: '',
                            status: 0,
                            text: 'menu item',
                            structure: {
                                classes: {
                                    type: 'text',
                                },
                                status: {
                                    type: 'text',
                                },
                                styles: {
                                    nested: [],
                                    type: 'json'
                                },
                                text: {
                                    type: 'text',
                                },
                            },
                            styles: [],
                        },
                        description: {
                            classes: '',
                            status: 0,
                            text: 'simple menu item',
                            structure: {
                                classes: {
                                    type: 'text',
                                },
                                status: {
                                    type: 'text',
                                },
                                styles: {
                                    nested: [],
                                    type: 'json'
                                },
                                text: {
                                    type: 'text',
                                },
                            },
                            styles: [],
                        },
                        styles: [],
                    };
                }
            }
        }
    }
</script>
