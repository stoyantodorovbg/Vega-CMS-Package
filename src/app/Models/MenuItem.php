<?php

namespace App\Models;

class MenuItem extends BasicModel
{
    /**
     * @var array
     */
    protected $guarded = [];

    /**
     * @var array
     */
    protected $casts = [
        'title' => 'array',
        'description' => 'array',
        'styles' => 'array'
    ];

    /**
     * The menu item belongs to a menu
     *
     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
     */
    public function menu()
    {
        return $this->belongsTo(Menu::class);
    }

    /**
     * The menu item may has a parent menu item
     *
     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
     */
    public function parentMenuItem()
    {
        return $this->belongsTo(MenuItem::class, 'parent_id');
    }

    /**
     * The menu item may has manu child menu items
     *
     * @return \Illuminate\Database\Eloquent\Relations\HasMany
     */
    public function childMenuItems()
    {
        return $this->hasMany(MenuItem::class, 'parent_id');
    }

    /**
     * Load all child menu items recursivelly
     *
     * @param MenuItem $menuItem
     * @return void
     */
    public function loadAllChildItems(MenuItem $menuItem): void
    {
        $menuItem->load('childMenuItems');
        foreach ($this->childMenuItems as $childMenuItem) {
            $childMenuItem->load('childMenuItems');

            if($childMenuItem->childMenuItems->count()) {
                $childMenuItem->loadAllChildItems($childMenuItem);
            }
        }
    }
}
