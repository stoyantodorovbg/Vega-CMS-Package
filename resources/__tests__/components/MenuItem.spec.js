import { mount } from '@vue/test-utils';
import MenuItem from "../../js/components/menu/MenuItem";

describe("MenuItem Component", () => {
  test('should mount without crashing', () => {
    const wrapper = mount(MenuItem);

    expect(wrapper).toMatchSnapshot();
  });
});
