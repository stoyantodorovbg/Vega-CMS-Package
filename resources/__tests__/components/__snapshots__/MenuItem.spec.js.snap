// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem Component should mount without crashing 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
